src/librustc/hir/mod.rs

```diff
@@ -643,6 +643,79 @@ impl Generics {
             self.params.iter().map(|p| p.span).collect::<Vec<Span>>().into()
         }
     }
+
+    /// Suggest restricting a type param with a new bound.
+    pub fn suggest_constraining_type_param(
```

Comment:
Diagnostics code really shouldn't be here; only data definitions and some utility functions should. I think this can remain in `error_reporting.rs` as a free function.